File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int cacheSize, vector<string> cities) {
    int answer = 0;
    
    //캐시 크기가 0
    if(cacheSize == 0){
        answer = cities.size()*5;
        return answer;
    }
    
    
    vector<string> cache;
    for(int i=0; i<cities.size(); i++){
        string check = cities[i];
        transform(check.begin(), check.end(), check.begin(), ::tolower);    // 소문자 변환
        auto it = find(cache.begin(), cache.end(), check);
        //캐시에 있음
        if(it != cache.end()){
            cache.erase(it);
            cache.push_back(check);
            answer++;
            
        }
        //캐시에 없음
        else{
            //캐시에 빈자리 있음
            if(cache.size() < cacheSize)
                cache.push_back(check);
            
            //캐시에 빈자리 없음
            else{
                cache.erase(cache.begin()+0);
                cache.push_back(check);
            }
            answer+=5;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
def solution(cacheSize, cities):
    answer = 0
    
    # LRU 알고리즘
    # hit = 1 miss =5
    
    # 캐시 크기 0인경우
    if cacheSize == 0 : return len(cities)*5
    
    c_l = []
    
    for c in cities :
        c = c.lower()
        # hit
        if c in c_l :
            answer+=1
            c_l.remove(c)  # 현재 위치에서 제거
            c_l.append(c)  # 가장 최근 위치로 이동
                
        #miss
        else :
            answer+=5
            if cacheSize > len(c_l) :
                c_l.append(c)           
            elif c_l : 
                c_l.pop(0)
                c_l.append(c)


    return answer